File: backend/src/CmdFont.cpp
// CmdFont.cpp - This file is part of QPlot

/* QPlot - Publication quality 2D graphs with dual coordinate systems
   Copyright (C) 2014  Daniel Wagenaar
  
   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.
  
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
  
   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

// CmdFont.C

#include "CmdFont.h"


static CBuilder<CmdFont> cbFont("font");

bool CmdFont::usage() {
  return error("Usage: font family [bold] [italic] size");
}

bool CmdFont::parse(Statement const &s) {
  if (s.length()<3 || s.length()>5)
    return usage();
  if (s[1].typ!=Token::BAREWORD)
    return usage();
  if (s[s.length()-1].typ!=Token::NUMBER)
    return usage();
  for (int k=2; k<s.length()-1; k++) {
    if (s[k].typ==Token::BAREWORD) {
      if (s[k].str=="bold")
	continue;
      else if (s[k].str=="italic")
	continue;
    }
    return usage();
  }
  return true;
}


void CmdFont::render(Statement const &s, Figure &f, bool) {
  QFont font(s[1].str);
  //  font.setPointSizeF(pt2iu(s[s.length()-1].num));
  font.setPixelSize(pt2iu(s[s.length()-1].num));
  for (int k=2; k<s.length()-1; k++) {
    if (s[k].str=="bold")
      font.setWeight(QFont::Bold);
    else if (s[k].str=="italic")
      font.setItalic(true);
  }
  f.painter().setFont(font);
}

File: backend/python/cmdfont.py
# cmdfont.py - This file is part of QPlot

# QPlot - Publication quality 2D graphs with dual coordinate systems
# Copyright (C) 2014  Daniel Wagenaar
# (same GPL licence as original)

from __future__ import annotations

from typing import TYPE_CHECKING

from PyQt6.QtGui import QFont

from .command import Commandfrom .token_ import Tokenfrom .factor import pt2iu
if TYPE_CHECKING:
    from .statement import Statement    from .figure import Figure

@Command.register("font")
class CmdFont(Command):
    """Set the current font.

    Syntax:
        font family [bold] [italic] size_pt
    """

    def _usage(self) -> bool:
        return self.error("Usage: font family [bold] [italic] size")

    def parse(self, s: Statement) -> bool:
        if not (3 <= len(s) <= 5):
            return self._usage()
        if s[1].typ != Token.BAREWORD:
            return self._usage()
        if s[len(s) - 1].typ != Token.NUMBER:
            return self._usage()
        for k in range(2, len(s) - 1):
            if s[k].typ == Token.BAREWORD and s[k].str in ("bold", "italic"):
                continue
            return self._usage()
        return True

    def render(self, s: Statement, f: Figure, dryrun: bool) -> None:
        font = QFont(s[1].str)
        font.setPixelSize(int(pt2iu(s[len(s) - 1].num)))
        for k in range(2, len(s) - 1):
            if s[k].str == "bold":
                font.setWeight(QFont.Weight.Bold)
            elif s[k].str == "italic":
                font.setItalic(True)
        f.painter().setFont(font)
